### PR TITLE
crimson/seastore: add Extentmap node type in cache.cc

### PR DESCRIPTION
**Pull request by liu-chunmei**
and check node size before make  merge.

Signed-off-by: chunmei-liu <chunmei.liu@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by liu-chunmei** [comment]
@tchaikov @athanatos  add Extent map node type in alloc_new_extent_by_type() and fix a bug. please review.

File: src/crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node_impl.cc [hunk 1]
```diff
@@ -73,7 +73,7 @@ ExtMapInnerNode::rm_lextent(ext_context_t ec, objaddr_t lo, lext_map_val_t val)
   auto rm_pt = get_containing_child(lo);
   return extmap_load_extent(ec, rm_pt->get_val(),  get_meta().depth - 1).safe_then(
     [this, ec, rm_pt, lo, val=std::move(val)](auto extent) mutable {
-    if (extent->at_min_capacity()) {
+    if (extent->at_min_capacity() && get_node_size() != 1) {
```
Comment (tchaikov) [inline src/crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node_impl.cc:76]: could you move this change into a separate commit? also, does it make sense to change like
```c++
if (extent->at_min_capacity() && get_node_size() > 1) {
```

to help reader to understand that we need another entry in the internal node to from which the pointers/elements are transferred.

File: src/crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node.h [hunk 2]
```diff
@@ -81,6 +81,7 @@ struct ExtMapNode : LogicalCachedExtent {
 
   virtual bool at_max_capacity() const = 0;
   virtual bool at_min_capacity() const = 0;
+  virtual uint16_t get_node_size() const = 0;
```
Comment (athanatos) [inline src/crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node.h:84]: Just make this unsigned.

**Comment by athanatos** [comment]
LGTM other than the comments.

**Comment by liu-chunmei** [comment]
@tchaikov updated. please review again.

**Review by tchaikov — APPROVED** [review]

**Review by athanatos — APPROVED** [review]

**Comment by liu-chunmei** [comment]
@tchaikov @athanatos add internal node test case.  please review again.

File: src/test/crimson/seastore/test_extmap_manager.cc [hunk 3]
```diff
@@ -242,39 +257,83 @@ TEST_F(extentmap_manager_test_t, force_split_balanced)
...
+    auto t = tm->create_transaction();
+    int i = 0;
+    for (auto iter = test_ext_mappings.begin(); iter != test_ext_mappings.end();) {
+      auto [lo, ext] = *iter;
+      ++iter;
+      rm_extent(extmap_root, *t, lo, ext);
+      i++;
+
+      if (i % 10 == 0) {
+        logger().error("submitting transaction i= {}", i);
```
Comment (tchaikov) [inline src/test/crimson/seastore/test_extmap_manager.cc:274]: please don't use `error()` for informative messages.

Comment (liu-chunmei) [reply src/test/crimson/seastore/test_extmap_manager.cc]: updated.

**Review by tchaikov — CHANGES_REQUESTED** [review]
```
207 - unittest_extmap_manager (Child aborted)
```

**Comment by liu-chunmei** [comment]
> ```
> 207 - unittest_extmap_manager (Child aborted)
> ```

@tchaikov  could you post the crash dump?  on my machine, it works.  I can't see any error.   ( it will last over 10 minutes. since the last test case need trigger internal node split and merge.)

**Comment by tchaikov** [comment]
please follow the link of our "make check" run, or https://jenkins.ceph.com/job/ceph-pull-requests/62705/console .

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by athanatos** [comment]
Ah, it's a side effect of another PR that just merged changing the init transaction to be weak.  I'll have a look.

**Comment by liu-chunmei** [comment]
@tchaikov Sam said: https://github.com/ceph/ceph/pull/37907 will solve the problem.

**Comment by tchaikov** [comment]
jenkins test make check

**Review by tchaikov — COMMENTED** [review]
```
208/209 Test #207: unittest_extmap_manager ...................   Passed  1249.27 sec
```

it takes more than 20 minutes to run `unittest_extmap_manager`. is it possible to take less time to run the test and with a relatively decent coverage?

**Comment by liu-chunmei** [comment]
> ```
> 208/209 Test #207: unittest_extmap_manager ...................   Passed  1249.27 sec
> ```
> 
> it takes more than 20 minutes to run `unittest_extmap_manager`. is it possible to take less time to run the test and with a relatively decent coverage?

I can remove the last test case which will take over 10mins, that case test internal node spilt and merge.  But I think in our seastore case, we will never use extentmap tree internal node split and merge, since each object not too big.

**Comment by liu-chunmei** [comment]
@tchaikov I removed the last test case, then current only need 1~2mins to run this test.

**Comment by tchaikov** [comment]
> @tchaikov I removed the last test case, then current only need 1~2mins to run this test.

could disable the time-consuming tests by following https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#temporarily-disabling-tests . probably you could have this in a follow-up test.

**Review by tchaikov — APPROVED** [review]
